File: skills/code-architect/references/code-examples/python/memento.py
"""Memento — capture and restore an object's state without breaking encapsulation."""

from __future__ import annotations


class _Memento:
    """Opaque snapshot; only the originator interprets its contents."""

    def __init__(self, state: str) -> None:
        self._state = state


class Editor:
    """Originator that produces and consumes mementos."""

    def __init__(self) -> None:
        self._content = ""

    def type(self, text: str) -> None:
        self._content += text

    def save(self) -> _Memento:
        return _Memento(self._content)

    def restore(self, memento: _Memento) -> None:
        self._content = memento._state

    def __str__(self) -> str:
        return self._content


class History:
    """Caretaker: stores mementos without inspecting them."""

    def __init__(self) -> None:
        self._stack: list[_Memento] = []

    def push(self, memento: _Memento) -> None:
        self._stack.append(memento)

    def pop(self) -> _Memento | None:
        return self._stack.pop() if self._stack else None


if __name__ == "__main__":
    editor, history = Editor(), History()
    editor.type("hello ")
    history.push(editor.save())
    editor.type("world")
    print("before undo:", editor)
    snapshot = history.pop()
    if snapshot:
        editor.restore(snapshot)
    print("after undo: ", editor)

File: skills/code-architect/references/code-examples/typescript/Memento.ts
// Memento — capture and restore an object's state without breaking encapsulation.

// Opaque snapshot; only the originator understands its contents.
class Memento {
  constructor(private readonly state: string) {}
  // Restricted accessor used solely by the originator.
  getState(): string {
    return this.state;
  }
}

class TextEditor {
  private content = "";

  type(text: string): void {
    this.content += text;
  }

  save(): Memento {
    return new Memento(this.content);
  }

  restore(memento: Memento): void {
    this.content = memento.getState();
  }

  get text(): string {
    return this.content;
  }
}

// Caretaker keeps mementos but never inspects them.
class History {
  private stack: Memento[] = [];
  backup(memento: Memento): void {
    this.stack.push(memento);
  }
  pop(): Memento | undefined {
    return this.stack.pop();
  }
}

function demo(): void {
  const editor = new TextEditor();
  const history = new History();

  editor.type("Hello");
  history.backup(editor.save());
  editor.type(", world");
  console.log(`before undo: "${editor.text}"`);

  const snapshot = history.pop();
  if (snapshot) editor.restore(snapshot);
  console.log(`after undo:  "${editor.text}"`);
}

demo();
